### ISSUE TITLE
Show error messages in result array for failed bulk inserts via HTTP endpoint

### ISSUE THREAD

**Issue by ckurze** [crate/crate#16649]
### Problem Statement

We are inserting many rows via python, i.e. the HTTP interface of CrateDB.

[Since version 4.2](https://github.com/crate/crate/blob/54105287c92be84a29d223e18a7e29ef676e7ae3/docs/appendices/release-notes/4.2.0.rst?plain=1#L53-L56), it reports the rowcount in the results array, either 1 for success, or -2 in case of failure and does not show an error.

In order to avoid single inserts to be executed to get the particular error message for the failed inserts as outlined in https://github.com/crate/cratedb-toolkit/pull/262, it would be great to see the particular error code and exception message directly in the results array (not a full stack trace as the [error_trace=true](https://cratedb.com/docs/crate/reference/en/latest/interfaces/http.html#error-handling) parameter would enable).

**Current output:**
```
{
    "cols": [],
    "duration": 31.547579,
    "results": [
        {
            "rowcount": 1
        },
        {
            "rowcount": -2
        },
        {
            "rowcount": -2
        }
    ]
}
```

**Ideal output:**
```
{
    "cols": [],
    "duration": 31.547579,
    "results": [
        {
            "rowcount": 1
        },
        {
            "rowcount": -2,
            "error": 4091,
            "message": "DuplicateKeyException[A document with the same primary key exists already]"
        },
        {
            "rowcount": -2,
            "error": 4000,
            "message": "SQLParseException[Cannot cast object element `a` with value `asdf` to type `bigint`]"
        }
    ]
}
```

Data to reproduce the ideally expected result above:

```
CREATE TABLE insert_test (id TEXT PRIMARY KEY, val OBJECT(DYNAMIC));

-- seed table
INSERT INTO insert_test(id, val) VALUES ('1', '{ "a": 1, "b": { "c": 1}, "d": [{"k": "k1", "v": 1}]}');

-- insert works, no error to report
INSERT INTO insert_test(id, val) VALUES ('2', '{ "a": 2, "b": { "c": 2}, "d": [{"k": "k1", "v": 2}]}');

-- DuplicateKeyException[A document with the same primary key exists already]
INSERT INTO insert_test(id, val) VALUES ('2', '{ "a": 2, "b": { "c": 2}, "d": [{"k": "k1", "v": 2}]}');

-- SQLParseException[Cannot cast object element `a` with value `asdf` to type `bigint`]
INSERT INTO insert_test(id, val) VALUES ('3', '{ "a": "asdf", "b": { "c": 3}, "d": [{"k": "k1", "v": 3}]}');
```

HTTP bulk insert of the above statements:
```
curl -sS -H 'Content-Type: application/json' \
-X POST '127.0.0.1:4200_sql' -d@- <<- EOF
{
  "stmt": "INSERT INTO insert_test(id, val) VALUES (?, ?)",
  "bulk_args": [
    ["2", "{ \"a\": 2, \"b\": { \"c\": 2}, \"d\": [{\"k\": \"k1\", \"v\": 2}]}"],
    ["2", "{ \"a\": 2, \"b\": { \"c\": 2}, \"d\": [{\"k\": \"k1\", \"v\": 2}]}"],
    ["3", "{ \"a\": \"asdf\", \"b\": { \"c\": 3}, \"d\": [{\"k\": \"k1\", \"v\": 3}]}"]
  ]
}
EOF
```

### Possible Solutions

If we don't want to change the default behaviour, maybe a switch to enable error reporting can be introduced.

### Considered Alternatives

An alternative would be to execute single inserts for each failed record as done in https://github.com/crate/cratedb-toolkit/pull/262, but when processing a lot of records that are failing, this leads to many single inserts significantly reducing the performance when required to be executed for 1000's of records.

**Comment by ckurze** [comment]
Note: When preparing and testing the sample data, I ran into https://github.com/crate/crate/issues/16617#issuecomment-2365175867

**Comment by proddata** [comment]
I’m wondering if it’s necessary to always return the full message, or if just returning the error code would be sufficient. For programmatic handling of ingest errors, the code alone should be enough?

Additionally, should either of these options be configurable through query parameters, similar to `?error_trace=true `or `?types` I’m concerned about large payloads when inserting 10k, 20k, or 50k records and huge returned messages.

**Comment by romseygeek** [comment]
Does it depend on the type of error?  In the "ideal output" section above, the DuplicateKeyException is going to be identical every time it is encountered, so the description is unnecessary, but having the specific data that caused an error in the SQLParseException case I can see be being useful.

**Comment by mfussenegger** [comment]
> I’m concerned about large payloads when inserting 10k, 20k, or 50k records and huge returned messages.

Afaik the current plan is to limit the amount of messages that get returned.
If we introduce another setting the main question to me would be how users are expected to use it.

E.g. for crash we have a `\verbose`. 
But how would users of DB APIs like python/crate-python set this?


I'd rather leave it out for now and return the messages by default (with limit). We can still follow up with configuration options later.

**Comment by matriv** [comment]
As a user of a bulk API, I would start checking the first error message (or the first few error messages), fix the issue, retry, check again and so on. So I believe that just limiting the errors (with messages) is enough for a user, if you return (just an example) 10k errors, and of them 1000 have full message and the rest only the error code, I don't believe it adds some value to the user that would at first spend time to resolve the first 1000 errors.

**Comment by seut** [comment]
We've implemented this by #16799 and choose to stop showing error details once more than `10` errors happen on a shard (avoid too much memory consumption by error handling). The exact limit to reach depends on how much shards are involved in a bulk operation. 
If this turns out to not be sufficient for a concrete use-case, we can iterate on this once the related non-satisfying use-case is known.
